activerecord/lib/active_record/connection_adapters/abstract/transaction.rb

```diff
@@ -542,7 +542,7 @@ def within_new_transaction(isolation: nil, joinable: true)
           ensure
             unless error
               if Thread.current.status == "aborting"
-                rollback_transaction
+                rollback_transaction if @connection.active?
```

Comment:
My first reaction is that this skips way too much -- record callbacks, likely leaves this object's transaction state out of sync, etc

Reply:
That's fair. I don't know how other adapters behave when the encounter a timeout, so this is tough. I'll try to dig further into this method and see if I can conditionalize at the right level. Thanks!

Reply:
I moved this line down to the actual SQL executing line: `connection.rollback_db_transaction`